### PR TITLE
Annotate activation functions with Nx blocks

### PR DESCRIPTION
**Pull request by Chapaman**
## Summary

- Continues [elixir-nx/axon#636](https://github.com/elixir-nx/axon/pull/636)

### CODE REVIEW

File: lib/axon/activations.ex [hunk 1]
```diff
@@ -517,13 +517,19 @@ defmodule Axon.Activations do
...
-      Nx.max(x, 0),
+      relu_block(x),
       [x],
       fn g -> [Nx.select(Nx.greater(x, 0), g, Nx.broadcast(0, g))] end
     )
   end
 
+  defblock Relu, relu_block(x) do
+    Nx.max(x, 0)
+  end
```
Comment (polvalente) [inline lib/axon/activations.ex]: I'm not sure that custom_grad should sit outside of Nx.block. The grad of block is the grad of the default callback IIRC, so I don't think this should matter

Comment (Chapaman) [reply lib/axon/activations.ex]: > I'm not sure that custom_grad should sit outside of Nx.block. The grad of block is the grad of the default callback IIRC, so I don't think this should matter

You're right on the default callback, The thing I'm having trouble with is on forwarding. The callback gets re-run on real tensors and `custom_grad`/`stop_grad` turns results into a traced tensor

Comment (polvalente) [reply lib/axon/activations.ex]: `custom_grad` and `stop_grad` should work though. Maybe `defblock` is not using proper `defn` scoping for the callback body?

Comment (Chapaman) [reply lib/axon/activations.ex]: maybe we should make sure that `custom_grad`and `stop_grad` are guarded on Nx?

File: lib/axon/activations.ex [hunk 2]
```diff
@@ -686,54 +676,43 @@ defmodule Axon.Activations do
...
-      |> reciprocal()
-      |> Nx.multiply(stable_exp)
-
-    # Cache logits so they are available in certain calculations,
-    # e.g. binary_cross_entropy and categorical_cross_entropy
-    cache_logits(x, res)
+    stable_exp
+    |> Nx.sum(axes: axes, keep_axes: true)
+    |> reciprocal()
+    |> Nx.multiply(stable_exp)
```
Comment (polvalente) [inline lib/axon/activations.ex:715]: cache_logits seems to be missing from this implementation

File: lib/axon/activations.ex [hunk 3]
```diff
@@ -686,54 +676,43 @@ defmodule Axon.Activations do
...
       >
 
   """
   defn softmax(x, opts \\ []) do
+    # Logits metadata must wrap the block result (not sit inside it) so
+    # losses can pattern-match `%{logits: _}` on the returned tensor.
+    cache_logits(x, softmax_block(x, opts))
+  end
+
+  defblock SoftMax, softmax_block(x, opts \\ []) do
```
Comment (polvalente) [inline lib/axon/activations.ex]: Maybe this warrants `defblockp` where the inner deftransform is also defined with `deftransformp` so that this module just advertises `softmax` but not `softmax_block`

File: lib/axon/block.ex [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      end
+
+    quote do
+      defmodule unquote(struct_module) do
+        @moduledoc false
+        unquote(struct_def)
+      end
+
+      # Public transform first so a preceding @doc attaches here, not to defnp.
+      deftransform unquote(name)(unquote_splicing(args)) do
```
Comment (polvalente) [inline lib/axon/block.ex]:
```suggestion
     unquote(kind)(unquote(name)(unquote_splicing(args))) do
```

where `kind` is either `:deftransform` or `:deftransformp` coming from `defblock` and `defblockp` respectively, I think is the correct solution to my other comment in activations

**Review by polvalente — APPROVED** [review]

**Review by polvalente — APPROVED** [review]
